[Security] Add a normalization step for the user-identifier in firewalls

Spomky opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | yes
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #49269
| License       | MIT
| Doc PR        | **TODO**

With this PR, the user identifier is lowered and normalized to make sure they are case-insensitive.
When enable, please note that the user provider or the user repository should be adapt:

* user identifier stored in the future shall be lowered and normalized as well
* the queries for existing user should be adapt

This is a very first attempt to resolve issue #49269. Any comments and reactions are welcome.

-----

See also https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#user-ids

nicolas-grekas commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php:
```diff
@@ -49,19 +53,26 @@ class UserBadge implements BadgeInterface
      * $userIdentifier as username.
      */
-    public function __construct(string $userIdentifier, callable $userLoader = null, array $attributes = null)
+    public function __construct(string $userIdentifier, callable $userLoader = null, array $attributes = null, bool $normalizeIdentifier = null)
```
I'd suggest accepting a closure instead, which would be used as the service doing the normalization, allowing to use whatever logic for that + pass it to all places that require the same processing.

Spomky replied:
Indeed. I will update this PR and propose a callable instead.

Spomky replied:
👌🏼 The PR is now up to date and accepts a callable.

Spomky commented:
Rebased to target Symfony 7.1

nicolas-grekas commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php:
```diff
@@ -36,6 +36,9 @@ class UserBadge implements BadgeInterface
     private ?array $attributes;
 
+    /** @var callable */
+    private $identifierNormalizer;
```
```suggestion
    private ?\Closure $identifierNormalizer = null;
```

nicolas-grekas commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php:
```diff
@@ -49,20 +52,20 @@ class UserBadge implements BadgeInterface
         }
-
+        $this->identifierNormalizer = $identifierNormalizer ?? fn ($userIdentifier) => $userIdentifier;
         $this->userIdentifier = $userIdentifier;
```
```suggestion
        if ($identifierNormalizer) {
            $this->identifierNormalizer = static fn () => $identifierNormalizer($userIdentifier);
        } else {
            $this->userIdentifier = $userIdentifier;
        }
        ;
```

nicolas-grekas commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php:
```diff
@@ -49,20 +52,20 @@ class UserBadge implements BadgeInterface
     public function getUserIdentifier(): string
     {
-        return $this->userIdentifier;
+        return ($this->identifierNormalizer)($this->userIdentifier);
```
trying to provide memoization:
```suggestion
        $userIdentifier = $this->userIdentifier ??= ($this->identifierNormalizer)();
        $this->identifierNormalizer = null;
        
        return $userIdentifier;
```

alternative we allow userIdentifier to be a closure and when that's the case, we call it to resolve it to a string

Spomky replied:
Many thanks.
PR up to date.

nicolas-grekas reviewed (commented):
needs a changelog entry :)

stloyd commented on src/Symfony/Component/Security/Http/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+7.1
+---
+
+ * Add argument `$identifierNormalizer` to `UseBadge::__construct()` to allow normalizing the username
```
```suggestion
 * Add argument `$identifierNormalizer` to `UserBadge::__construct()` to allow normalizing the identifier
```

Spomky replied:
Many thanks!

nicolas-grekas commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php:
```diff
@@ -49,19 +51,29 @@ class UserBadge implements BadgeInterface
+            $this->identifierNormalizer = static fn () => $identifierNormalizer($userIdentifier);
+        } else {
+            $this->userIdentifier = $userIdentifier;
+        }
```
```suggestion
        }

```

nicolas-grekas commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php:
```diff
@@ -49,19 +51,29 @@ class UserBadge implements BadgeInterface
     {
+        if (isset($this->userIdentifier)) {
+            return $this->userIdentifier;
+
+        }
```
```suggestion
        }

```

nicolas-grekas commented on src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php:
```diff
@@ -36,6 +36,8 @@ class UserBadge implements BadgeInterface
     private UserInterface $user;
     private ?array $attributes;
 
+    private ?\Closure $identifierNormalizer;
```
```suggestion
    private ?\Closure $identifierNormalizer = null;
```

Spomky commented:
Comments addressed, PR rebased and squashed.
Ready for the next step

nicolas-grekas commented on src/Symfony/Component/Security/Http/Tests/Authenticator/Passport/Badge/UserBadgeTest.php:
```diff
@@ -23,4 +27,29 @@ public function testUserNotFound()
+        static::assertSame($expectedNormalizedIdentifier, $badge->getUserIdentifier());
+    }
+
+    public function provideUserIdentifierNormalizationData(): iterable
```
```suggestion
    public static function provideUserIdentifierNormalizationData(): iterable
```

nicolas-grekas commented:
>  When enable, please note that the user provider or the user repository should be adapt:

do we have any advice on the topic? Or any things we can anticipate that could be part of the same PR?

Spomky commented:
> do we have any advice on the topic? Or any things we can anticipate that could be part of the same PR?

I checked how the built-in user providers could handle this, but I'm not sure if any of them can be changed as well.
How the user ID is stored or retrieved is highly dependent on the developer's implementation.
To me, the most important now seems to propagate this into the documentation.

nicolas-grekas commented:
What's the next step here? How do we make it easy to opt-in for the OWASP-recommended normalization?

Spomky commented:
> What's the next step here? How do we make it easy to opt-in for the OWASP-recommended normalization?

Hello @nicolas-grekas,

This PR is now up to date, rebased against the 7.3 branch, and squashed. It’s ready for review.

The next step is to finalize the documentation. It should demonstrate how to inject a normalizer to ensure that different variations (e.g., `JOHN.DOE`, `john.doe`, `John.DOe`) are treated as equivalent.

BR.

nicolas-grekas commented:
Can you give us a hint about how this can be used? : )

Spomky commented:
This feature will most likely be used by the firewall Authenticator service that creates the UserBadge, but it could also be used by the `TokenHandler::getUserBadge` when handling access tokens.

Below is a simple example of how to implement a `NormalizedUserBadge`.
In this example, it ensures that any variations in the username—such as uppercase, lowercase, or accented characters—are normalized to a consistent format.

```php
namespace App\Security;

use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
use Symfony\Component\String\UnicodeString;
use function Symfony\Component\String\u;

final class NormalizedUserBadge extends UserBadge
{
    public function __construct(string $identifie)
    {
        $callback = static fn (string $identifier) => u($identifier)->normalize(UnicodeString::NFKC)->ascii()->lower()->toString();
        
        parent::__construct($identifier, identifierNormalizer: $callback);
    }
}
```

From the Authenticator:

```php
<?php

namespace App\Security;

final class PasswordAuthenticator extends AbstractLoginFormAuthenticator
{
    // Simplified for brievety
    public function authenticate(Request $request): Passport
    {
        $username = (string) $request->request->get('username', '');
        $password = (string) $request->request->get('password', '');

        $request->getSession()
            ->set(SecurityRequestAttributes::LAST_USERNAME, $username);

        return new Passport(
            new NormalizedUserBadge($username),
            new PasswordCredentials($password),
            [
                //All other useful badges
            ]
        );
    }
```

When a user attempts to log in with “John.Doe” or “JOhn.doe,” both will be treated as “john.doe.”
Keep in mind that any existing usernames in the database should also be normalized accordingly.

Another potential use case involves scenarios where users can log in with various forms of the same identifier: `john.doe@acme.com`, `acme.com\jdoe`, `https://amce.com/+jdoe` or `acct:jdoe@acme.com`.
By using this feature, you can unify all these variants into a single, standardized format (e.g., “domain\nickname”) without having to create custom queries on the repository side.
In this case, the callback could be as follows (not tested):

```php
        $callback = static function (string $identifier) {
            $normalized = u($identifier)->normalize(UnicodeString::NFKC)->ascii()->lower()->toString();

            $username = '';
            $domain = '';

            match (true) {
                str_contains($normalized, '@') => [
                    $localPart, $domain = explode('@', $normalized, 2),
                    $username = str_contains($localPart, '.')
                        ? substr(explode('.', $localPart, 2)[0], 0, 1) . explode('.', $localPart, 2)[1]
                        : $localPart,
                ],
                str_contains($normalized, '\\') => [
                    $domain, $username = explode('\\', $normalized, 2)
                ],
                str_contains($normalized, '://') => [
                    $parts = parse_url($normalized),
                    isset($parts['host'], $parts['path']) && [
                        $domain = $parts['host'],
                        $username = ltrim($parts['path'], '/')
                    ]
                ],
                default => $username = $normalized
            };

            $username = preg_replace('/[^a-z0-9]/', '', $username);

            return $domain . '\\' . $username;
        };
```

Spomky committed:
Add a normalization step for the user-identifier in firewalls

welcoMattic reviewed (approved):
Thanks for this @Spomky! 

I think the documentation of this feature must provide real life example to help users to use it properly (I think mostly about username policy of Google where `foo.bar@gmail.com`, `foobar@gmail.com`, and `f.o.o.b.a.r@gmail.com` refers to the same user).

For the code, it's a 👍 for me

fabpot commented:
Thank you @Spomky.

fabpot commented:
As a next step, I'm wondering if we should somehow define some reusable normalizers that developers can use without reinventing the wheel (like the Google strategy mentioned above).
